### ISSUE TITLE
你好

### ISSUE THREAD

**Issue by dengyufei123** [koalaman/shellcheck#2762]
#### For bugs
- Rule Id (if any, e.g. SC1000): 
- My shellcheck version (`shellcheck --version` or 'online'): online
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

#### For new checks and feature suggestions
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [ ] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/your/shebang
Your script
```

#### Here's what shellcheck currently says:
```
TODO: Describe current output
```
#### Here's what I wanted or expected to see:

TODO: Describe expected/desired output

**Comment by dengyufei123** [comment]
#!/bin/sh

export LANG=zh_CN.UTF-8
REPORT_FILE="/deng"
DATE=DATE=$(date +%Y-%m-%d_%H.%M.%S)


#一级菜单
menu1()
{
        clear
        cat <<EOF
----------------------------------------
|****   欢迎使用cetnos7.9优化脚本    ****|
|****      博客地址: aaa.al         ****|
----------------------------------------
1. 日常巡检
2. 进行系统加固
3. 完成必要的数据备份
4. 一键自动部署lnmp环境
5. 手动选者是否安装wordpress
6. 退出
EOF
        read -p "please enter your choice[1-6]:" num1
}



menu2()
{
 clear
 cat <<EOF
        ----------------------------------------
        |****Please Enter Your Choice:[0-26]****|
        ----------------------------------------
        1.系统检查
        2.CPU检查
        3.内存检查
        4.磁盘检查
        5.网络检查
        6.监听检查
        7.进程服务
        8.服务检查
        9.自启动检查
        10.登录检查
        11.计划任务检查
        12.用户检查
        13.密码检查
        14.Sudoers检查
        15.JDK检查
        16.防火墙检查
        17.SSH检查
        18.syslog检查
        19.SNMP检查
        20.NTP检查
        21.软件检查
        22.查看所有被修改过的文件返回最近24小时内的
        23.检查定时文件的完整性
        24.查看系统命令是否被替换
        25.返回上一级
        26.退出
EOF
 read -p "please enter your choice[1-26]:" num2
 
}
menu3()
{
 clear
 cat <<EOF
----------------------------------------
|****Please Enter Your Choice:[0-13]****|
----------------------------------------
1. 修改字符集
2. 关闭selinux
3. 关闭firewalld
4. 精简开机启动
5. 修改文件描述符
6. 安装常用工具及修改yum源
7. 优化系统内核
8. 加快ssh登录速度
9. 禁用ctrl+alt+del重启
10.设置时间同步
11.history优化
12.返回上级菜单
13.退出
EOF
 read -p "please enter your choice[1-13]:" num3
 
}

#1.系统检查
xitong()
{
 echo "########## 系统检查 ##########" >> $REPORT_FILE
 echo "$(cat /etc/redhat-release)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE 
}

#2.CPU检查
cpu() 
{
 echo "########## CPU ##########" >> $REPORT_FILE
 echo "$(lscpu)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}

#3.内存检查
neicun()
{
 echo "########## 内存检查 ##########" >> $REPORT_FILE
 echo "$(free -h)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}

#4.磁盘检查
cipan()
{
 echo "########## 磁盘检查 ##########" >> $REPORT_FILE
 echo "$(df -h)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}

#5.网络检查
wangluo()
{
 echo "########## 网络检查 ##########" >> $REPORT_FILE
 echo "$(ifconfig)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}

#6.监听检查
jianting()
{
 echo "########## 监听检查 ##########" >> $REPORT_FILE
 echo "$(netstat -ant)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}

#7.进程检查
jincheng()
{
 echo "########## 进程检查 ##########" >> $REPORT_FILE
 echo "$(ps aux)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}

#8.服务检查
fuwu()
{
 echo "########## 服务检查 ##########" >> $REPORT_FILE
 echo "$(service --status-all)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}

#9.自启动检查
ziqidong()
{
 echo "########## 自启动检查 ##########" >> $REPORT_FILE
 echo "$(systemctl list-unit-files | grep enabled)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}

#10.登录检查
denglu()
{
 echo "########## 登录检查 ##########" >> $REPORT_FILE
 echo "$(who)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}

#11.计划任务检查
jihuarenwu()
{
 echo "########## 计划任务检查 ##########" >> $REPORT_FILE
 echo "$(crontab -l)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}
#12.用户检查
yonghu()
{
 echo "########## 用户检查 ##########" >> $REPORT_FILE
 echo "$(cat /etc/passwd)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}
#13.密码检查
mima()
{
 echo "########## 密码检查 ##########" >> $REPORT_FILE
 echo "$(cat /etc/shadow)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}
#14.Sudoers检查
sudoers()
{
 echo "########## Sudoers检查 ##########" >> $REPORT_FILE
 echo "$(cat /etc/sudoers)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}
#15.JDK检查
jdk()
{
 echo "########## JDK检查 ##########" >> $REPORT_FILE
 echo "$(java -version)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}
#16.防火墙检查
fanghuoqiang()
{
 
}echo "########## 防火墙检查 ##########" >> $REPORT_FILE
 echo "$(iptables -L)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
#17.SSH检查
ssh()
{
 echo "########## SSH检查 ##########" >> $REPORT_FILE
 echo "$(cat /etc/ssh/sshd_config)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}

#18.syslog检查
syslog()
{
 echo "########## syslog检查 ##########" >> $REPORT_FILE
 echo "$(cat /var/log/syslog)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}
#19.SNMP检查
snmp()
{
 echo "########## SNMP检查 ##########" >> $REPORT_FILE
 echo "$(cat /etc/snmp/snmpd.conf)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}
#20.NTP检查
ntp()
{
 echo "########## NTP检查 ##########" >> $REPORT_FILE
 echo "$(cat /etc/ntp.conf)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}
#21.软件检查
ruanjian()
{
 echo "########## 软件检查 ##########" >> $REPORT_FILE
 echo "$(dpkg -l)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}
#22.查看所有被修改过的文件返回最近24小时内的
wenjian()
{
 echo "########## 查看所有被修改过的文件返回最近24小时内的 ##########" >> $REPORT_FILE
 echo "$(find / -mtime -1 -type f)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}
#23.检查定时文件的完整性
wenjianwanzheng()
{
 echo "########## 检查定时文件的完整性 ##########" >> $REPORT_FILE
 echo "$(diff /var/spool/cron/crontabs/root /etc/crontab)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}
#24.查看系统命令是否被替换
xitongtihuan()
{
 echo "########## 查看系统命令是否被替换 ##########" >> $REPORT_FILE
 echo "$(dpkg -S $(which ls) | grep -E -o '/.*:$')" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}

#1.修改字符集
localeset()
{
 echo "========================修改字符集========================="
 cat > /etc/locale.conf <<EOF
LANG="zh_CN.UTF-8"
#LANG="en_US.UTF-8"
SYSFONT="latarcyrheb-sun16"
EOF
 source /etc/locale.conf
 echo "#cat /etc/locale.conf"
 cat /etc/locale.conf
 action "完成修改字符集" /bin/true
 echo "==========================================================="
 sleep 2
}

#2.关闭selinux
selinuxset() 
{
 selinux_status=`grep "SELINUX=disabled" /etc/sysconfig/selinux | wc -l`
 echo "========================禁用SELINUX========================"
 if [ $selinux_status -eq 0 ];then
  sed  -i "s#SELINUX=enforcing#SELINUX=disabled#g" /etc/sysconfig/selinux
  setenforce 0
  echo '#grep SELINUX=disabled /etc/sysconfig/selinux'
  grep SELINUX=disabled /etc/sysconfig/selinux
  echo '#getenforce'
  getenforce
 else
  echo 'SELINUX已处于关闭状态'
  echo '#grep SELINUX=disabled /etc/sysconfig/selinux'
                grep SELINUX=disabled /etc/sysconfig/selinux
                echo '#getenforce'
                getenforce
 fi
  action "完成禁用SELINUX" /bin/true
 echo "==========================================================="
 sleep 2
}

#3.关闭firewalld
firewalldset()
{
 echo "=======================禁用firewalld========================"
 systemctl stop firewalld.service &> /dev/null
 echo '#firewall-cmd  --state'
 firewall-cmd  --state
 systemctl disable firewalld.service &> /dev/null
 echo '#systemctl list-unit-files | grep firewalld'
 systemctl list-unit-files | grep firewalld
 action "完成禁用firewalld，生产环境下建议启用！" /bin/true
 echo "==========================================================="
 sleep 5
}

#4.精简开机启动
chkset()
{
 echo "=======================精简开机启动========================"
 systemctl disable auditd.service
 systemctl disable postfix.service
 systemctl disable dbus-org.freedesktop.NetworkManager.service
 echo '#systemctl list-unit-files | grep -E "auditd|postfix|dbus-org\.freedesktop\.NetworkManager"'
 systemctl list-unit-files | grep -E "auditd|postfix|dbus-org\.freedesktop\.NetworkManager"
 action "完成精简开机启动" /bin/true
 echo "==========================================================="
 sleep 2
}

#5.修改文件描述符
limitset()
{
 echo "======================修改文件描述符======================="
 echo '* - nofile 65535'>/etc/security/limits.conf
 ulimit -SHn 65535
 echo "#cat /etc/security/limits.conf"
 cat /etc/security/limits.conf
 echo "#ulimit -Sn ; ulimit -Hn"
 ulimit -Sn ; ulimit -Hn
 action "完成修改文件描述符" /bin/true
 echo "==========================================================="
 sleep 2
}

#6.安装常用工具及修改yum源
yumset()
{
 echo "=================安装常用工具及修改yum源==================="
 yum install wget -y &> /dev/null
 if [ $? -eq 0 ];then
  cd /etc/yum.repos.d/
  \cp CentOS-Base.repo CentOS-Base.repo.$(date +%F)
  ping -c 1 mirrors.aliyun.com &> /dev/null
  if [ $? -eq 0 ];then
   wget -O /etc/yum.repos.d/CentOS-Base.repo http://mirrors.aliyun.com/repo/Centos-7.repo &> /dev/null
   yum clean all &> /dev/null
   yum makecache &> /dev/null
  else
   echo "无法连接网络"
       exit $?
    fi
 else
  echo "wget安装失败"
  exit $?
 fi
 yum -y install ntpdate lsof net-tools telnet vim lrzsz tree nmap nc sysstat &> /dev/null
 action "完成安装常用工具及修改yum源" /bin/true
 echo "==========================================================="
 sleep 2
}

#7. 优化系统内核
kernelset()
{
 echo "======================优化系统内核========================="
 chk_nf=`cat /etc/sysctl.conf | grep conntrack |wc -l`
 if [ $chk_nf -eq 0 ];then
  cat >>/etc/sysctl.conf<<EOF
net.ipv4.tcp_fin_timeout = 2
net.ipv4.tcp_tw_reuse = 1
net.ipv4.tcp_tw_recycle = 1
net.ipv4.tcp_syncookies = 1
net.ipv4.tcp_keepalive_time = 600
net.ipv4.ip_local_port_range = 4000 65000
net.ipv4.tcp_max_syn_backlog = 16384
net.ipv4.tcp_max_tw_buckets = 36000
net.ipv4.route.gc_timeout = 100
net.ipv4.tcp_syn_retries = 1
net.ipv4.tcp_synack_retries = 0
net.core.somaxconn = 16384
net.core.netdev_max_backlog = 16384
net.ipv4.tcp_max_orphans = 16384
net.netfilter.nf_conntrack_max = 25000000
net.netfilter.nf_conntrack_tcp_timeout_established = 180
net.netfilter.nf_conntrack_tcp_timeout_time_wait = 120
net.netfilter.nf_conntrack_tcp_timeout_close_wait = 60
net.netfilter.nf_conntrack_tcp_timeout_fin_wait = 120
EOF
 sysctl -p
 else
  echo "优化项已存在。"
 fi
 action "内核调优完成" /bin/true
 echo "==========================================================="
 sleep 2
}

#8.加快ssh登录速度
sshset()
{
 echo "======================加快ssh登录速度======================"
 sed -i 's#^GSSAPIAuthentication yes$#GSSAPIAuthentication no#g' /etc/ssh/sshd_config
 sed -i 's/#UseDNS yes/UseDNS no/g' /etc/ssh/sshd_config
 systemctl restart sshd.service
 echo "#grep GSSAPIAuthentication /etc/ssh/sshd_config"
 grep GSSAPIAuthentication /etc/ssh/sshd_config
 echo "#grep UseDNS /etc/ssh/sshd_config"
 grep UseDNS /etc/ssh/sshd_config
 action "完成加快ssh登录速度" /bin/true
 echo "==========================================================="
 sleep 2
}

#9. 禁用ctrl+alt+del重启
restartset()
{
 echo "===================禁用ctrl+alt+del重启===================="
 rm -rf /usr/lib/systemd/system/ctrl-alt-del.target
 action "完成禁用ctrl+alt+del重启" /bin/true
 echo "==========================================================="
 sleep 2
}

#10. 设置时间同步
ntpdateset()
{
 echo "=======================设置时间同步========================"
 yum -y install ntpdate &> /dev/null
 if [ $? -eq 0 ];then
  /usr/sbin/ntpdate time.windows.com
  echo "*/5 * * * * /usr/sbin/ntpdate ntp.aliyun.com &>/dev/null" >> /var/spool/cron/root
 else
  echo "ntpdate安装失败"
  exit $?
 fi
 action "完成设置时间同步" /bin/true
 echo "==========================================================="
 sleep 2
}

#11. history优化
historyset()
{
 echo "========================history优化========================"
 chk_his=`cat /etc/profile | grep HISTTIMEFORMAT |wc -l`
 if [ $chk_his -eq 0 ];then
  cat >> /etc/profile <<'EOF'
#设置history格式
export HISTTIMEFORMAT="[%Y-%m-%d %H:%M:%S] [`whoami`] [`who am i|awk '{print $NF}'|sed -r 's#[()]##g'`]: "
#记录shell执行的每一条命令
export PROMPT_COMMAND='\
if [ -z "$OLD_PWD" ];then
    export OLD_PWD=$PWD;
fi;
if [ ! -z "$LAST_CMD" ] && [ "$(history 1)" != "$LAST_CMD" ]; then
    logger -t `whoami`_shell_dir "[$OLD_PWD]$(history 1)";
fi;
export LAST_CMD="$(history 1)";
export OLD_PWD=$PWD;'
EOF
  source /etc/profile
 else
  echo "优化项已存在。"
 fi
 action "完成history优化" /bin/true
 echo "==========================================================="
 sleep 2
}

#控制函数
main()
{
 menu1
 case $num1 in
  1)
   menu2
   case $num2 in
            1)
            xitong
            ;;
            2)
            cpu
            ;;
            3)
            neicun
            ;;
            4)
            cipan
            ;;
            5)
            wangluo
            ;;
            6)
            jianting
            ;;
            7)
            jincheng
            ;;
            8)
            fuwu
            ;;
            9)
            ziqidong
            ;;
            10)
            denglu
            ;;
            11)
            jihuarenwu
            ;;
            12)
            yonghu
            ;;
            13)
            mima
            ;;
            14)
            sudoers
            ;;
            15)
            jdk
            ;;
            16)
            fanghuoqiang
            ;;
            17)
            ssh
            ;;
            18)
            syslog
            ;;
            19)
            snmp
            ;;
            20)
            ntp
            ;;
            21)
            ruanjian
            ;;
            22)
            wenjian
            ;;
            23)
            wenjianwanzheng
            ;;
            24)
            xitongtihuan
            ;;
            25)
            main
            ;;
            26)
            exit
            ;;
            *)
      echo 'Please select a number from [1-26].'
      ;;
   esac
   ;;
   2)
   menu3
   case $num3 in
                  1)
                    localeset
                    ;;
                  2)
                    selinuxset
                    ;;
                  3)
                    firewalldset
                    ;;
                  4)
                    chkset
                    ;;
                  5)
                    limitset
                    ;;
                  6)     
            yumset
                    ;;
                  7)
                    kernelset
                    ;;
                  8)
                    sshset
                    ;;
                  9)
                    restartset
                    ;;
                  10)
                    ntpdateset
                    ;;
    11)
      historyset
      ;;
    12)
      main
      ;;
    13)
      exit
      ;;
    *)
      echo 'Please select a number from [1-13].'
      ;;
   esac
   ;;
  
  6)
   exit
   ;;
  *)
   echo 'Err:Please select a number from [1-6].'
   sleep 3
   main
   ;;
 esac
}
main $*
export LANG=zh_CN.UTF-8
REPORT_FILE="/deng"
DATE=DATE=$(date +%Y-%m-%d_%H.%M.%S)


#一级菜单
menu1()
{
        clear
        cat <<EOF
----------------------------------------
|****   欢迎使用cetnos7.9优化脚本    ****|
|****      博客地址: aaa.al         ****|
----------------------------------------
1. 日常巡检
2. 进行系统加固
3. 完成必要的数据备份
4. 一键自动部署lnmp环境
5. 手动选者是否安装wordpress
6. 退出
EOF
        read -p "please enter your choice[1-6]:" num1
}



menu2()
{
 clear
 cat <<EOF
        ----------------------------------------
        |****Please Enter Your Choice:[0-26]****|
        ----------------------------------------
        1.系统检查
        2.CPU检查
        3.内存检查
        4.磁盘检查
        5.网络检查
        6.监听检查
        7.进程服务
        8.服务检查
        9.自启动检查
        10.登录检查
        11.计划任务检查
        12.用户检查
        13.密码检查
        14.Sudoers检查
        15.JDK检查
        16.防火墙检查
        17.SSH检查
        18.syslog检查
        19.SNMP检查
        20.NTP检查
        21.软件检查
        22.查看所有被修改过的文件返回最近24小时内的
        23.检查定时文件的完整性
        24.查看系统命令是否被替换
        25.返回上一级
        26.退出
EOF
 read -p "please enter your choice[1-26]:" num2
 
}
menu3()
{
 clear
 cat <<EOF
----------------------------------------
|****Please Enter Your Choice:[0-13]****|
----------------------------------------
1. 修改字符集
2. 关闭selinux
3. 关闭firewalld
4. 精简开机启动
5. 修改文件描述符
6. 安装常用工具及修改yum源
7. 优化系统内核
8. 加快ssh登录速度
9. 禁用ctrl+alt+del重启
10.设置时间同步
11.history优化
12.返回上级菜单
13.退出
EOF
 read -p "please enter your choice[1-13]:" num3
 
}

#1.系统检查
xitong()
{
 echo "########## 系统检查 ##########" >> $REPORT_FILE
 echo "$(cat /etc/redhat-release)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE 
}

#2.CPU检查
cpu() 
{
 echo "########## CPU ##########" >> $REPORT_FILE
 echo "$(lscpu)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}

#3.内存检查
neicun()
{
 echo "########## 内存检查 ##########" >> $REPORT_FILE
 echo "$(free -h)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}

#4.磁盘检查
cipan()
{
 echo "########## 磁盘检查 ##########" >> $REPORT_FILE
 echo "$(df -h)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}

#5.网络检查
wangluo()
{
 echo "########## 网络检查 ##########" >> $REPORT_FILE
 echo "$(ifconfig)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}

#6.监听检查
jianting()
{
 echo "########## 监听检查 ##########" >> $REPORT_FILE
 echo "$(netstat -ant)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}

#7.进程检查
jincheng()
{
 echo "########## 进程检查 ##########" >> $REPORT_FILE
 echo "$(ps aux)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}

#8.服务检查
fuwu()
{
 echo "########## 服务检查 ##########" >> $REPORT_FILE
 echo "$(service --status-all)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}

#9.自启动检查
ziqidong()
{
 echo "########## 自启动检查 ##########" >> $REPORT_FILE
 echo "$(systemctl list-unit-files | grep enabled)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}

#10.登录检查
denglu()
{
 echo "########## 登录检查 ##########" >> $REPORT_FILE
 echo "$(who)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}

#11.计划任务检查
jihuarenwu()
{
 echo "########## 计划任务检查 ##########" >> $REPORT_FILE
 echo "$(crontab -l)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}
#12.用户检查
yonghu()
{
 echo "########## 用户检查 ##########" >> $REPORT_FILE
 echo "$(cat /etc/passwd)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}
#13.密码检查
mima()
{
 echo "########## 密码检查 ##########" >> $REPORT_FILE
 echo "$(cat /etc/shadow)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}
#14.Sudoers检查
sudoers()
{
 echo "########## Sudoers检查 ##########" >> $REPORT_FILE
 echo "$(cat /etc/sudoers)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}
#15.JDK检查
jdk()
{
 echo "########## JDK检查 ##########" >> $REPORT_FILE
 echo "$(java -version)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}
#16.防火墙检查
fanghuoqiang()
{
 
}echo "########## 防火墙检查 ##########" >> $REPORT_FILE
 echo "$(iptables -L)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
#17.SSH检查
ssh()
{
 echo "########## SSH检查 ##########" >> $REPORT_FILE
 echo "$(cat /etc/ssh/sshd_config)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}

#18.syslog检查
syslog()
{
 echo "########## syslog检查 ##########" >> $REPORT_FILE
 echo "$(cat /var/log/syslog)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}
#19.SNMP检查
snmp()
{
 echo "########## SNMP检查 ##########" >> $REPORT_FILE
 echo "$(cat /etc/snmp/snmpd.conf)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}
#20.NTP检查
ntp()
{
 echo "########## NTP检查 ##########" >> $REPORT_FILE
 echo "$(cat /etc/ntp.conf)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}
#21.软件检查
ruanjian()
{
 echo "########## 软件检查 ##########" >> $REPORT_FILE
 echo "$(dpkg -l)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}
#22.查看所有被修改过的文件返回最近24小时内的
wenjian()
{
 echo "########## 查看所有被修改过的文件返回最近24小时内的 ##########" >> $REPORT_FILE
 echo "$(find / -mtime -1 -type f)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}
#23.检查定时文件的完整性
wenjianwanzheng()
{
 echo "########## 检查定时文件的完整性 ##########" >> $REPORT_FILE
 echo "$(diff /var/spool/cron/crontabs/root /etc/crontab)" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}
#24.查看系统命令是否被替换
xitongtihuan()
{
 echo "########## 查看系统命令是否被替换 ##########" >> $REPORT_FILE
 echo "$(dpkg -S $(which ls) | grep -E -o '/.*:$')" >> $REPORT_FILE
 echo "" >> $REPORT_FILE
}

#1.修改字符集
localeset()
{
 echo "========================修改字符集========================="
 cat > /etc/locale.conf <<EOF
LANG="zh_CN.UTF-8"
#LANG="en_US.UTF-8"
SYSFONT="latarcyrheb-sun16"
EOF
 source /etc/locale.conf
 echo "#cat /etc/locale.conf"
 cat /etc/locale.conf
 action "完成修改字符集" /bin/true
 echo "==========================================================="
 sleep 2
}

#2.关闭selinux
selinuxset() 
{
 selinux_status=`grep "SELINUX=disabled" /etc/sysconfig/selinux | wc -l`
 echo "========================禁用SELINUX========================"
 if [ $selinux_status -eq 0 ];then
  sed  -i "s#SELINUX=enforcing#SELINUX=disabled#g" /etc/sysconfig/selinux
  setenforce 0
  echo '#grep SELINUX=disabled /etc/sysconfig/selinux'
  grep SELINUX=disabled /etc/sysconfig/selinux
  echo '#getenforce'
  getenforce
 else
  echo 'SELINUX已处于关闭状态'
  echo '#grep SELINUX=disabled /etc/sysconfig/selinux'
                grep SELINUX=disabled /etc/sysconfig/selinux
                echo '#getenforce'
                getenforce
 fi
  action "完成禁用SELINUX" /bin/true
 echo "==========================================================="
 sleep 2
}

#3.关闭firewalld
firewalldset()
{
 echo "=======================禁用firewalld========================"
 systemctl stop firewalld.service &> /dev/null
 echo '#firewall-cmd  --state'
 firewall-cmd  --state
 systemctl disable firewalld.service &> /dev/null
 echo '#systemctl list-unit-files | grep firewalld'
 systemctl list-unit-files | grep firewalld
 action "完成禁用firewalld，生产环境下建议启用！" /bin/true
 echo "==========================================================="
 sleep 5
}

#4.精简开机启动
chkset()
{
 echo "=======================精简开机启动========================"
 systemctl disable auditd.service
 systemctl disable postfix.service
 systemctl disable dbus-org.freedesktop.NetworkManager.service
 echo '#systemctl list-unit-files | grep -E "auditd|postfix|dbus-org\.freedesktop\.NetworkManager"'
 systemctl list-unit-files | grep -E "auditd|postfix|dbus-org\.freedesktop\.NetworkManager"
 action "完成精简开机启动" /bin/true
 echo "==========================================================="
 sleep 2
}

#5.修改文件描述符
limitset()
{
 echo "======================修改文件描述符======================="
 echo '* - nofile 65535'>/etc/security/limits.conf
 ulimit -SHn 65535
 echo "#cat /etc/security/limits.conf"
 cat /etc/security/limits.conf
 echo "#ulimit -Sn ; ulimit -Hn"
 ulimit -Sn ; ulimit -Hn
 action "完成修改文件描述符" /bin/true
 echo "==========================================================="
 sleep 2
}

#6.安装常用工具及修改yum源
yumset()
{
 echo "=================安装常用工具及修改yum源==================="
 yum install wget -y &> /dev/null
 if [ $? -eq 0 ];then
  cd /etc/yum.repos.d/
  \cp CentOS-Base.repo CentOS-Base.repo.$(date +%F)
  ping -c 1 mirrors.aliyun.com &> /dev/null
  if [ $? -eq 0 ];then
   wget -O /etc/yum.repos.d/CentOS-Base.repo http://mirrors.aliyun.com/repo/Centos-7.repo &> /dev/null
   yum clean all &> /dev/null
   yum makecache &> /dev/null
  else
   echo "无法连接网络"
       exit $?
    fi
 else
  echo "wget安装失败"
  exit $?
 fi
 yum -y install ntpdate lsof net-tools telnet vim lrzsz tree nmap nc sysstat &> /dev/null
 action "完成安装常用工具及修改yum源" /bin/true
 echo "==========================================================="
 sleep 2
}

#7. 优化系统内核
kernelset()
{
 echo "======================优化系统内核========================="
 chk_nf=`cat /etc/sysctl.conf | grep conntrack |wc -l`
 if [ $chk_nf -eq 0 ];then
  cat >>/etc/sysctl.conf<<EOF
net.ipv4.tcp_fin_timeout = 2
net.ipv4.tcp_tw_reuse = 1
net.ipv4.tcp_tw_recycle = 1
net.ipv4.tcp_syncookies = 1
net.ipv4.tcp_keepalive_time = 600
net.ipv4.ip_local_port_range = 4000 65000
net.ipv4.tcp_max_syn_backlog = 16384
net.ipv4.tcp_max_tw_buckets = 36000
net.ipv4.route.gc_timeout = 100
net.ipv4.tcp_syn_retries = 1
net.ipv4.tcp_synack_retries = 0
net.core.somaxconn = 16384
net.core.netdev_max_backlog = 16384
net.ipv4.tcp_max_orphans = 16384
net.netfilter.nf_conntrack_max = 25000000
net.netfilter.nf_conntrack_tcp_timeout_established = 180
net.netfilter.nf_conntrack_tcp_timeout_time_wait = 120
net.netfilter.nf_conntrack_tcp_timeout_close_wait = 60
net.netfilter.nf_conntrack_tcp_timeout_fin_wait = 120
EOF
 sysctl -p
 else
  echo "优化项已存在。"
 fi
 action "内核调优完成" /bin/true
 echo "==========================================================="
 sleep 2
}

#8.加快ssh登录速度
sshset()
{
 echo "======================加快ssh登录速度======================"
 sed -i 's#^GSSAPIAuthentication yes$#GSSAPIAuthentication no#g' /etc/ssh/sshd_config
 sed -i 's/#UseDNS yes/UseDNS no/g' /etc/ssh/sshd_config
 systemctl restart sshd.service
 echo "#grep GSSAPIAuthentication /etc/ssh/sshd_config"
 grep GSSAPIAuthentication /etc/ssh/sshd_config
 echo "#grep UseDNS /etc/ssh/sshd_config"
 grep UseDNS /etc/ssh/sshd_config
 action "完成加快ssh登录速度" /bin/true
 echo "==========================================================="
 sleep 2
}

#9. 禁用ctrl+alt+del重启
restartset()
{
 echo "===================禁用ctrl+alt+del重启===================="
 rm -rf /usr/lib/systemd/system/ctrl-alt-del.target
 action "完成禁用ctrl+alt+del重启" /bin/true
 echo "==========================================================="
 sleep 2
}

#10. 设置时间同步
ntpdateset()
{
 echo "=======================设置时间同步========================"
 yum -y install ntpdate &> /dev/null
 if [ $? -eq 0 ];then
  /usr/sbin/ntpdate time.windows.com
  echo "*/5 * * * * /usr/sbin/ntpdate ntp.aliyun.com &>/dev/null" >> /var/spool/cron/root
 else
  echo "ntpdate安装失败"
  exit $?
 fi
 action "完成设置时间同步" /bin/true
 echo "==========================================================="
 sleep 2
}

#11. history优化
historyset()
{
 echo "========================history优化========================"
 chk_his=`cat /etc/profile | grep HISTTIMEFORMAT |wc -l`
 if [ $chk_his -eq 0 ];then
  cat >> /etc/profile <<'EOF'
#设置history格式
export HISTTIMEFORMAT="[%Y-%m-%d %H:%M:%S] [`whoami`] [`who am i|awk '{print $NF}'|sed -r 's#[()]##g'`]: "
#记录shell执行的每一条命令
export PROMPT_COMMAND='\
if [ -z "$OLD_PWD" ];then
    export OLD_PWD=$PWD;
fi;
if [ ! -z "$LAST_CMD" ] && [ "$(history 1)" != "$LAST_CMD" ]; then
    logger -t `whoami`_shell_dir "[$OLD_PWD]$(history 1)";
fi;
export LAST_CMD="$(history 1)";
export OLD_PWD=$PWD;'
EOF
  source /etc/profile
 else
  echo "优化项已存在。"
 fi
 action "完成history优化" /bin/true
 echo "==========================================================="
 sleep 2
}

#控制函数
main()
{
 menu1
 case $num1 in
  1)
   menu2
   case $num2 in
            1)
            xitong
            ;;
            2)
            cpu
            ;;
            3)
            neicun
            ;;
            4)
            cipan
            ;;
            5)
            wangluo
            ;;
            6)
            jianting
            ;;
            7)
            jincheng
            ;;
            8)
            fuwu
            ;;
            9)
            ziqidong
            ;;
            10)
            denglu
            ;;
            11)
            jihuarenwu
            ;;
            12)
            yonghu
            ;;
            13)
            mima
            ;;
            14)
            sudoers
            ;;
            15)
            jdk
            ;;
            16)
            fanghuoqiang
            ;;
            17)
            ssh
            ;;
            18)
            syslog
            ;;
            19)
            snmp
            ;;
            20)
            ntp
            ;;
            21)
            ruanjian
            ;;
            22)
            wenjian
            ;;
            23)
            wenjianwanzheng
            ;;
            24)
            xitongtihuan
            ;;
            25)
            main
            ;;
            26)
            exit
            ;;
            *)
      echo 'Please select a number from [1-26].'
      ;;
   esac
   ;;
   2)
   menu3
   case $num3 in
                  1)
                    localeset
                    ;;
                  2)
                    selinuxset
                    ;;
                  3)
                    firewalldset
                    ;;
                  4)
                    chkset
                    ;;
                  5)
                    limitset
                    ;;
                  6)     
            yumset
                    ;;
                  7)
                    kernelset
                    ;;
                  8)
                    sshset
                    ;;
                  9)
                    restartset
                    ;;
                  10)
                    ntpdateset
                    ;;
    11)
      historyset
      ;;
    12)
      main
      ;;
    13)
      exit
      ;;
    *)
      echo 'Please select a number from [1-13].'
      ;;
   esac
   ;;
  
  6)
   exit
   ;;
  *)
   echo 'Err:Please select a number from [1-6].'
   sleep 3
   main
   ;;
 esac
}
main $*

**Comment by brother** [comment]
@dengyufei123 I have removed your comment. 1278 lines of code is abit over the top.

Make a smaller code block to demonstrate the problem if needed. And add context,.
